Custom WebView2Loader implementation

SteffenL opened the pull request:
As we do not have access to the source code of Microsoft's WebView2Loader library, it causes several annoying issues:

* MinGW-w64/GCC can only link it dynamically while Visual C++ can also link it statically.
* Go/cgo uses MinGW-w64/GCC which excludes the option to link it statically.
* Developers must manually manage the WebView2 dependency and make sure to bundle and ship the correct DLLs.
* Go/cgo tooling does not currently support a workflow that makes all of this automatic for those who use our Go bindings. Instead they are by default met with an error due the WebView2 header not being found if the environment is not set up according to our instructions (note 1 and 3).
* It is not obvious how to set up an IDE to work on one's own project when using our Go bindings due to the WebView2 dependency.

This PR attempts to reimplement parts of the WebView2Loader in order to eliminate the need for the binaries. We still need the header (note 2 and 3). After starting a reverse engineering effort I figured that `WebView2Loader.dll` is not actually doing a whole lot and merely provides some convenience for finding and using the client DLL of the browser (`EmbeddedBrowserWebView.dll`). 

I found [OpenWebView2Loader](https://github.com/jchv/OpenWebView2Loader) which I then relied on as my RE effort became more or less redundant. One can simply drop in this open-source implementation and it works just fine. We do not need the exact behavior of the official WebView2Loader which is why I did not do this.

This work is not yet complete but it is functional and I think it is a good time to begin discussions.

This branch has instructions (e.g. URLs) changed specifically for this work in order to test the work before a merge:

https://github.com/SteffenL/webview/tree/testing/custom-webview2-loader-impl

## Noteworthy Changes

- Minimalistic reimplementation of the features we need from WebView2Loader.
- If the official loader DLL (`WebView2Loader.dll`) is present at runtime then it will be explicitly linked at runtime and used instead of our custom implementation. This also eliminates the need for an import library (`*.lib` file). See note 4.
- Readme has been updated to reflect that our loader implementation will be used when the official one is not present. The WebView2 SDK requirement is about the same as before.

## Notes

1. Go has code generation support but it requires users to manually run `go generate` before building their programs. As an experiment I wrote a tool for fetching WebView2 and setting up the Go environment. This tool can be invoked via `go generate` by adding `//go:generate go run github.com/webview/[...]` to the go source file. It works well but it is not a natural workflow for Go.
2. Technically we can generate the header file of WebView2 by compiling the `WebView2.idl` file using the [MIDL Compiler](https://docs.microsoft.com/en-us/windows/win32/com/midl-compiler). That works fine but requires Visual Studio/C++ tools. It would be nice to be able to get [WIDL](https://www.winehq.org/docs/widl) (Wine Interface Definition Language (IDL) compiler) to generate a working header as it is part of MinGW-w64 but the generated code gave me compile errors.
3. WebView2 SDK  can be pulled in and used by Go bindings using vendoring. To reduce the scope of this PR this is not included in this PR anymore.
4. To get this exact behavior, projects using this library should stop unnecessarily linking the loader's import library (`*.lib` file). Failing to do so should however produce the same result because the linker (at least in case of MSVC) sees that no exported symbols are used.

## To Do

* [x] `CreateCoreWebView2EnvironmentWithOptions`
* [x] `GetAvailableCoreWebView2BrowserVersionString`
* [x] Browser version check.
* [x] Support WebView2 runtimes installed system-wide.
* [x] Support WebView2 runtimes installed per-user.
* [x] Custom browser executable folder.
* [x] Retry logic for WebView2 initialization.

## Won't Do

* Overrides for options via environment variables, e.g. `WEBVIEW2_BROWSER_EXECUTABLE_FOLDER`.
* Support Microsoft Edge Insider (preview) release channels (Beta, Dev, Canary).
* Non-essential `WebView2Loader.dll` exports.

dandeto commented:
WebView2Loader.lib is the static library. MinGW can't link that? That's a bummer. Many projects ship with dlls without a fuss, but I understand the desire to ditch them, of course. It's still possible to compile with a static lib using the windows compiler, but not for Go. It looks like supporting Go and other bindings is the main issue here, correct? They will need to depend upon a dynamic library without the option of a static one?

Replacing the WebView2Loader with the opensource alternative is something I will look into a little more. I think most people will want the official one just for the sake of it's status and stability, but I could see those who are affected by the MinGW linking deficiency wanting this option to be available.

SteffenL commented:
The naming style of the lib files is a bit unusual:

* `WebView2Loader.dll.lib` - shared.
* `WebView2LoaderStatic.lib` - static.

Unfortunately, the static one is not usable by MinGW-w64/GCC. Static/Dynamic linking should always be an option in my opinion but users of Go and MinGW-w64 users do no thave that choice.

Much of the issue has to do with Go/cgo since it uses MinGW-w64 and I believe it breaks the normal worklow of a Go developer if it does not by default link statically. It is not limited to Go though since C/C++ users have the option to use MinGW-w64 instead of Visual C++ unlike before. Other bindings would also be able to benefit from this. I personally build the library mainly using MinGW-w64 now in order to have a consistent workflow across platforms.

Instead of fully replacing WebView2Loader with the open-source alternative I would rather reimplement only the minimum amount of features needed to ensure that we can maintain runtime stability and good quality code. It also uses WRL which is is something we eliminated recently to be more compatible with MinGW-w64 and old versions of Windows, so I don't think we should reintroduce that.

For example I do not think that we need to be able to select at runtime which release channel to use of the browser engine, or to have the option to use an embedded browser or installed browser.

I think that the loader could be considered as a middleman that can be eliminated without any real consequences.

Having the option to use the loader anyway is something I would support and then comes the question about what should be the default behavior. The only behavior that gives Go users the natural workflow they expect is to get rid the need for them to deal with cgo.

SteffenL commented:
Commit 0f256cf9861bc880b272fe09c8446bb926a3c643 prefers `WebView2Loader.dll` if present. I tested mismatching versions: `WebView2.h` version `1.0.1150.38`; `WebView2Loader.dll` version `1.0.1245.22`. I think this is a supported use case while we need a version check for the other way around to be safe.

SteffenL commented:
Regarding the embedded WebView2 header, alternatives I can imagine for that right now is to either split the Go bindings from the main Git repository (preferable in my opinion), or if possible move the WebView2 header into a Git submodule if that is something Go/cgo can work with.

I am open to different ideas.

justjosias commented:
For historical reasons, I don't think we can split the Go bindings. Do you know of a way we can do that without breaking existing builds?

SteffenL commented:
I imagine that existing users of this repo should be locked to a specific version/commit so it would only affect new dependents or those who try to upgrade.

I have not tested this but I imagine we could redirect those who continue using this repo, e.g. something like this:

```
replace github.com/webview/webview v1.2.3 => github.com/webview/webview-go v1.2.3
```

SteffenL commented:
I have not gotten my idea above with `replace` to work so I have given up on that.

After quite a journey with trial and error, I believe that I have figured out an acceptable way to split the Go bindings from the main repository.

Since Go freezes versions then I believe that we have the option to remove the bindings from the main repository. Google caches the Go module so that it should keep working for current dependents even if we were to delete the whole repository. They just cannot upgrade to a new version.

We can also either strip down the Go module in the main repository or keep it as-is but issue a compile warning or error via cgo using `#warning` or `#error` since we are using cgo anyway:

```
//#warning The webview Go bindings have moved to https://github.com/webview/webview-go - please see the project readme.
import "C"
```

This way people should be able to "upgrade" but at least get this message after doing so for as long as they are using the old Go module.

MS WebView2 can be moved into a separate Git repository that we have to maintain. If we use my custom implementation of the loader then we do not need to host the lib/DLL files.

All of this will be tied together using so-called vendoring. Since we specify Go 1.13 as the minimum version in our `go.mod` file then we and dependents must add the flag `-mod=vendor` to the `go build` command. That flag became default since Go 1.14 so if we bump the minimum Go version we require to 1.14 then vendoring will apply automatically and will be more convenient for everyone who can use Go 1.14 and up.

What vendoring does is copy Go packages from dependencies into a `vendor` directory in the project directory and the mentioned flag makes Go consider the modules there when building instead of the local module cache (which will have unpredictable file paths).

In order to get all the non-Go sources we need into the vendor directory then those files **must** be in a directory that has Go sources. We can add dummy Go packages in each directory that we need to have in the `vendor` directory. That's just a one-line `*.go` file. This is an approach other people have taken. There is a long discussion at https://github.com/golang/go/issues/26366 closed with no official solution. Unfortunately this means that Go will pollute the main repository.

This way we are able to specify predictable include paths to the C/C++ compiler such as the following:

```
#cgo CFLAGS: -Ivendor/github.com/webview/webview -I../../../github.com/webview/webview
#cgo CXXFLAGS: -Ivendor/github.com/webview/webview -Ivendor/github.com/webview/mswebview2 -I../../../github.com/webview/webview-nogo -I../../../github.com/webview/mswebview2
```

`-Ivendor/...` would be for the current project to find its dependencies, and `-I../...` would be for dependencies to find their dependencies.

Dependents must be made aware of vendoring and run `go mod vendor` at least once before `go build`.

SteffenL commented:
This PR has now been configured for vendoring. The purpose of this PR is not splitting the Go bindings so I have left that for another time.

I have a complete example at the following locations:

* https://github.com/SteffenL/mswebview2
* https://github.com/SteffenL/webview-nogo
* https://github.com/SteffenL/webview-go
* https://github.com/SteffenL/webview-go-example

You can try it easily like this:

```sh
git clone https://github.com/SteffenL/webview-go-example
cd webview-go-example
go mod vendor
go build
# Go <1.14: go build -mod=vendor
webview-go-example.exe
```

SteffenL commented:
This should answer questions regarding API and version compatibility:

https://docs.microsoft.com/en-us/microsoft-edge/webview2/concepts/versioning#minimum-version-and-build-number-to-instantiate-webview2

> The WebView2 release SDK has been forward-compatible ever since version 1 (that is, SDK version 1.0.622.22).


> When you use a WebView2 SDK that has a build number equal to or less than the WebView2 Runtime, every API that you have access to in that SDK works with that version of the Runtime.

> When you use a WebView2 SDK that has a build number greater than the WebView2 Runtime, the newer APIs' implementations aren't available in the Runtime.

> The client should have a version of the Runtime that supports all of the latest APIs that are in the SDK version you use to develop the app. For full support for the latest APIs in a release version of the SDK, the Runtime on the client must have a build number that's greater than or equal to the SDK build number.

SteffenL commented:
In order to reduce the scope of this PR, I have removed the Go vendoring setup. I do however think we should explore that idea some more. The issue description has been changed to reflect the latest changes.

I have also added support for the following:

* WebView2 runtimes installed per-user.
* Custom browser executable directories (embedded), i.e. the `browserExecutableFolder` parameter of `CreateCoreWebView2EnvironmentWithOptions`.
* An implementation of `GetAvailableCoreWebView2BrowserVersionString`.

SteffenL commented:
Retry logic has now been implemented for WebView2 initialization more or less per official recommendations:

> The app should retry CreateCoreWebView2Controller upon failure, unless the error code is HRESULT_FROM_WIN32(ERROR_INVALID_STATE). When the app retries CreateCoreWebView2Controller upon failure, it is recommended that the app restarts from creating a new WebView2 Environment. If a WebView2 Runtime update happens, the version associated with a WebView2 Environment may have been removed and causing the object to no longer work. Creating a new WebView2 Environment works since it uses the latest version.
>
> WebView creation fails with HRESULT_FROM_WIN32(ERROR_INVALID_STATE) if a running instance using the same user data folder exists, and the Environment objects have different EnvironmentOptions. For example, if a WebView was created with one language, an attempt to create a WebView with a different language using the same user data folder will fail.
>
> The creation will fail with E_ABORT if parentWindow is destroyed before the creation is finished. If this is caused by a call to DestroyWindow, the creation completed handler will be invoked before DestroyWindow returns, so you can use this to cancel creation and clean up resources synchronously when quitting a thread.

https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2environment?view=webview2-1.0.1150.38

SteffenL commented:
I consider this work to be more or less finished but it might be a good idea to create a new official release before merging this.

dandeto commented:
I will take some time this week to review this in more detail so I fully understand what's going on. This looks like a lot of hard work, kudos @SteffenL!

janh commented on webview.h:
```diff
@@ -1290,9 +1290,8 @@ inline std::wstring find_edge_webview_client_dll(unsigned int min_api_version) {
   client_dll_path += L"x86";
-#elif defined(_M_ARM64)
-  // TODO: Check if this is actually possible
-  client_dll_path += L"arm64";
```
Building is possible with LLVM MinGW, and the resulting binary runs. So there isn't actually any reason to remove this.

SteffenL replied:
@janh I removed this particular code because I had no way to validate it. I would like to make sure that it works before merging it.

SteffenL replied:
I have finally been able to validate this code on ARM64 hardware.

dandeto commented:
The only downside I see to this is the added complexity to the single-header application, but all of the upsides are worth it.

The benefits of static-linking are obvious, and the fact that Go benefits so much from it is additionally compelling. I think getting rid of `WebView2Loader.dll` is great, even though Webview2.h is still required. It's one less thing we need to worry about. I'm glad the code you wrote is backwards-compatible so that if our implementation ever breaks in the future, users can still link `WebView2Loader.dll`.

2 separate discussions worth having:
Splitting Go bindings
Potentially refactoring the single header library into maybe 2 or more headers

I support merging this. Do you have any additional thoughts @justjosias?

SteffenL commented:
@dandeto If we split the WebView2Loader reimplementation into a separate file, I wonder if we might have some issues with the header inclusion order since `webview.h` should probably not depend on the new file.

Since #889 introduced `WEBVIEW_MSWEBVIEW2_EXPLICIT_LINK`, it might make sense to introduce something like `WEBVIEW_MSWEBVIEW2_BUILTIN_IMPL` as well to conditionally enable the built-in implementation of WebView2Loader.

SteffenL commented:
The latest work disables the built-in WebView2 loader implementation by default because `WEBVIEW_MSWEBVIEW2_EXPLICIT_LINK` was introduced recently.

The following compile-time options can be used to change how the library integrates the WebView2 loader:

* `WEBVIEW_MSWEBVIEW2_EXPLICIT_LINK` - The library supports linking `WebView2Loader.dll` explicitly, avoiding the need for import libraries (`*.lib`).
* `WEBVIEW_MSWEBVIEW2_BUILTIN_IMPL` - Enables the built-in implementation of the WebView2 loader. Implies `WEBVIEW_MSWEBVIEW2_EXPLICIT_LINK` to maximize compatibility.

I would personally like the built-in implementation to be enabled by default, with the official one taking priority. Maybe we can invert the options so that those who do not want this can explicitly disable the features rather than explicitly enable them. This is up for discussion.

dandeto commented:
@SteffenL I agree with you - I think it's best to have the built-in implementation enabled by default so that we get all of the benefits by default. I think most users aren't going to need to use the official loader.

SteffenL commented:
All right. I think we can merge this first and then take care of the "inverted" options so that the built-in implementation is enabled by default. That will be the most beneficial approach for sure. Thank you for the review!

SteffenL commented:
We could also just keep it simple, remove the compile-time options and always compile the built-in implementation with explicit linking of the official one taking priority.

Existing users will not notice any difference and there are plans to introduce error codes at some point so that missing DLLs do not silently fail.

What do you think, @dandeto?

SteffenL commented:
The options were kept in the latest commit but defaults were changed:

* Enable built-in WebView2Loader implementation by default.
* Link `WebView2Loader.dll` explicitly by default only if the built-in implementation is enabled.

| `*_BUILTIN_IMPL` | `*_EXPLICIT_LINK` | Effect                                    |
|:----------------:|:-----------------:| ----------------------------------------- |
| (undefined)      | (undefined)       | Built-in impl ON, explicit link ON.       |
| 1                | (undefined)       | Same as above.                            |
| (undefined)      | 1                 | Same as above.                            |
| 0                | 0                 | Built-in impl OFF, explicit link OFF.     |
| 0                | 1                 | Built-in impl OFF, explicit link ON.      |
| 1                | 0                 | Error - needs `*_EXPLICIT_LINK=1`.        |
| 1                | 1                 | Built-in impl ON, explicit link ON.       |

SteffenL commented:
Given that I adressed @dandeto's comment about making the built-in loader the new default (as originally intended) after the PR was approved while also leaving room to disable the options, I am going to merge this so that we can move on. Feel free to bring any new concerns to my attention so that I can address those later.
